File: BilisPera/BilisPera/Product/AlertView/BPDatePickerView.m
//
//  BPDatePickerView.m
//  BilisPera
//
//  Created by BHJ on 2025/8/25.
//

#import "BPDatePickerView.h"

NS_ASSUME_NONNULL_BEGIN
@implementation BPDatePickerViewModel

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.contentHeight = kRatio(436);
    }
    return self;
}

@end

@interface BPDatePickerView ()<UIPickerViewDelegate,UIPickerViewDataSource>

@property (nonatomic, strong) UIPickerView *pickerView;
@property (nonatomic, strong) UILabel *firstLabel;
@property (nonatomic, strong) UILabel *secondLabel;
@property (nonatomic, strong) UILabel *thirdLabel;
@property (nonatomic, strong) UIStackView *stackView;

@property (nonatomic, assign) NSInteger selectedYear;
@property (nonatomic, assign) NSInteger selectedMonth;
@property (nonatomic, assign) NSInteger selectedDay;
@property (nonatomic, assign) NSInteger selectedComponent;
@property (nonatomic, strong) NSArray<NSNumber *> *months;
@property (nonatomic, strong) NSArray<NSNumber *> *days;
@property (nonatomic, strong) NSDateComponents *currentDateCom;

@end

@implementation BPDatePickerView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _currentDateCom = [[NSCalendar currentCalendar] components:NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay fromDate:[NSDate date]];
        [self setupUI];
        [self applyModel];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame model:(BPDatePickerViewModel *)model {
    self = [super initWithFrame:frame];
    if (self) {
        self.model = model;
        _currentDateCom = [[NSCalendar currentCalendar] components:NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay fromDate:[NSDate date]];
        [self setupUI];
        [self applyModel];
    }
    return self;
}

#pragma mark - UI Setup

- (void)setupUI {
    self.backgroundColor = [UIColor clearColor];
    // PickerView
    _pickerView = [[UIPickerView alloc] initWithFrame:CGRectMake(0,
                                                                 0,
                                                                 kScreenW,
                                                                 kRatio(200))];
    _pickerView.backgroundColor = [UIColor clearColor];
    _pickerView.delegate = self;
    _pickerView.dataSource = self;
    _pickerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_pickerView];
    
    // Labels
    _firstLabel = [self createLabelWithText:@"Day"];
    _secondLabel = [self createLabelWithText:@"Month"];
    _thirdLabel = [self createLabelWithText:@"Year"];
    
    // StackView
    _stackView = [[UIStackView alloc] init];
    _stackView.axis = UILayoutConstraintAxisHorizontal;
    _stackView.alignment = UIStackViewAlignmentCenter;
    _stackView.distribution = UIStackViewDistributionFillEqually;
    _stackView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [_stackView addArrangedSubview:_firstLabel];
    [_stackView addArrangedSubview:_secondLabel];
    [_stackView addArrangedSubview:_thirdLabel];
    
    [self addSubview:_stackView];
    [self setupConstraints];
}

- (UILabel *)createLabelWithText:(NSString *)text {
    UILabel *label = [[UILabel alloc] init];
    label.backgroundColor = [UIColor clearColor];
    label.textColor = [UIColor blackColor];
    label.font = [UIFont systemFontOfSize:16];
    label.text = text;
    label.textAlignment = NSTextAlignmentCenter;
    label.translatesAutoresizingMaskIntoConstraints = NO;
    return label;
}

- (void)setupConstraints {
    [NSLayoutConstraint activateConstraints:@[
        // PickerView constraints
        [self.pickerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
        [self.pickerView.topAnchor constraintEqualToAnchor:self.topAnchor constant:kRatio(44)],
        [self.pickerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [self.pickerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
        [self.pickerView.heightAnchor constraintEqualToConstant:kRatio(436)],
        
        // StackView constraints
        [self.stackView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
        [self.stackView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [self.stackView.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.stackView.heightAnchor constraintEqualToConstant:kRatio(44)]
    ]];
}

#pragma mark - Model Application

- (void)applyModel {
    [self.pickerView reloadAllComponents];
    NSDate *currentDate = [NSDate dateFromString:self.model.currentDate withFormat:@"dd-MM-yyyy"] ?: [NSDate date];
    
    NSDateComponents *dateCom = [[NSCalendar currentCalendar] components:NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay fromDate:currentDate];
    
    NSInteger currentYear = dateCom.year ?: 0;
    NSInteger currentMonth = dateCom.month ?: 0;
    NSInteger currentDay = dateCom.day ?: 0;
    NSInteger currentYearIndex = self.currentDateCom.year - currentYear;
    
    [self.pickerView selectRow:currentYearIndex inComponent:2 animated:YES];
    [self.pickerView selectRow:currentMonth - 1 inComponent:1 animated:YES];
    [self.pickerView selectRow:currentDay - 1 inComponent:0 animated:YES];
    
    [self pickerView:self.pickerView didSelectRow:currentDay - 1 inComponent:0];
    [self pickerView:self.pickerView didSelectRow:currentMonth - 1 inComponent:1];
    [self pickerView:self.pickerView didSelectRow:currentYearIndex inComponent:2];
}

- (void)setModel:(BPDatePickerViewModel *)model{
    _model = model;
    [self applyModel];
}

#pragma mark - UIPickerViewDataSource

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 3;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    if (component == 2) {
        return 100; // Years
    } else if (component == 1) {
        return self.months.count;
    } else {
        return self.days.count;
    }
}

#pragma mark - UIPickerViewDelegate

- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component {
    return kRatio(50);
}

- (nullable NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    if (component == 2) {
        return [NSString stringWithFormat:@"%ld",
                (self.currentDateCom.year - row)];
    } else if (component == 1) {
        return [NSString stringWithFormat:@"%ld", (row + 1)];
    } else {
        return [NSString stringWithFormat:@"%ld", (row + 1)];
    }
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    self.selectedComponent = component;
    self.selectedYear = (self.currentDateCom.year) - [pickerView selectedRowInComponent:2];
    self.selectedMonth = [pickerView selectedRowInComponent:1] + 1;
    self.selectedDay = [pickerView selectedRowInComponent:0] + 1;
    
    NSString *dateString = [NSString stringWithFormat:@"%02ld-%02ld-%02ld",
                            self.selectedDay,
                            self.selectedMonth,
                            self.selectedYear];
    
    if (self.model.valueChanged) {
        self.model.valueChanged(dateString);
    }
    if (component == 2) {
        NSInteger year = [pickerView selectedRowInComponent:2] + self.currentDateCom.year;
        NSInteger currentMonth = [[NSDate date] currentMonth];
        NSInteger currentYear = [[NSDate date] currentYear];
        if (year == currentYear) {
            self.months = [self createMonthArrayUpTo:currentMonth];
        } else {
            self.months = [self createMonthArray];
        }
        [pickerView reloadComponent:1];
    }
    if (component == 1) {
        NSInteger year = [pickerView selectedRowInComponent:2] + self.currentDateCom.year;
        NSInteger month = [pickerView selectedRowInComponent:1] + 1;
        NSInteger daysInMonth = [self daysInMonth:month year:year];
        NSInteger currentMonth = [[NSDate date] currentMonth];
        NSInteger currentYear = [[NSDate date] currentYear];
        NSInteger currentDay = [[NSDate date] currentDay];
        if (year == currentYear && month == currentMonth) {
            self.days = [self createDayArrayUpTo:currentDay];
        } else {
            self.days = [self createDayArrayUpTo:daysInMonth];
        }
        [pickerView reloadComponent:0];
    }
}

- (NSArray<NSNumber *> *)createMonthArray {
    NSMutableArray *months = [NSMutableArray array];
    for (int i = 1; i <= 12; i++) {
        [months addObject:@(i)];
    }
    return [months copy];
}

- (NSArray<NSNumber *> *)createMonthArrayUpTo:(NSInteger)month {
    NSMutableArray *months = [NSMutableArray array];
    for (int i = 1; i <= month; i++) {
        [months addObject:@(i)];
    }
    return [months copy];
}

- (NSArray<NSNumber *> *)createDayArray {
    return [self createDayArrayUpTo:31];
}

- (NSArray<NSNumber *> *)createDayArrayUpTo:(NSInteger)day {
    NSMutableArray *days = [NSMutableArray array];
    for (int i = 1; i <= day; i++) {
        [days addObject:@(i)];
    }
    return [days copy];
}

- (NSInteger)daysInMonth:(NSInteger)month year:(NSInteger)year {
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *components = [[NSDateComponents alloc] init];
    [components setYear:year];
    [components setMonth:month];
    NSDate *date = [calendar dateFromComponents:components];
    NSRange range = [calendar rangeOfUnit:NSCalendarUnitDay inUnit:NSCalendarUnitMonth forDate:date];
    return range.length;
}

@end


NS_ASSUME_NONNULL_END

File: BilisPera/BilisPera/Product/AlertView/BPDatePickerView.swift
//
//  BPDatePickerView.swift
//  BilisPera
//
//  Created by BHJ on 2025/8/19.
//

import UIKit

class BPDatePickerView: UIView {
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        applyModel()
    }
    
    init(frame: CGRect,model: BPDatePickerViewModel = .init()) {
        self.model = model
        super.init(frame: frame)
        setupUI()
        applyModel()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc var model: BPDatePickerViewModel = .init(){
        didSet{
            applyModel()
        }
    }
    
    lazy var pickerView : UIPickerView = {
        let view = UIPickerView.init(
            frame: CGRect.init(x: 0, y: 0, width: kScreenW, height: 200.ratio())
        )
        view.backgroundColor = UIColor.clear
        view.delegate = self
        view.dataSource = self
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    lazy var firstLabel : UILabel = {
        let view = UILabel.init()
        view.backgroundColor = UIColor.clear
        view.textColor = UIColor.black
        view.font = UIFont.systemFont(ofSize: 16)
        view.text = "Day"
        view.textAlignment = .center
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    lazy var secondLabel : UILabel = {
        let view = UILabel.init()
        view.backgroundColor = UIColor.clear
        view.textColor = UIColor.black
        view.font = UIFont.systemFont(ofSize: 16)
        view.text = "Month"
        view.textAlignment = .center
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    lazy var thirdLabel : UILabel = {
        let view = UILabel.init()
        view.backgroundColor = UIColor.clear
        view.textColor = UIColor.black
        view.font = UIFont.systemFont(ofSize: 16)
        view.text = "Year"
        view.textAlignment = .center
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    lazy var stackView: UIStackView = {
        let view = UIStackView()
        view.axis = .horizontal
        view.alignment = .center
        view.distribution = .fillEqually
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    private var currentDateCom: DateComponents = Calendar.current.dateComponents(
        [.year, .month, .day],
        from: Date()
    )    //日期类型
    private var selectedYear: Int = 0
    private var selectedMonth: Int = 0
    private var selectedDay: Int = 0
    private var selectedComponent: Int = 0
    private var months: [Int] = Array(1...12)
    private var days: [Int] = Array(1...31)
}

extension BPDatePickerView: UIPickerViewDelegate,UIPickerViewDataSource{
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 3
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        if component == 2 {
            return 100
        } else if component == 1 {
            return months.count
        } else {
            return days.count
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 33.ratio()
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        if component == 2 {
            return "\((currentDateCom.year!) - row)"
        } else if component == 1 {
            return "\(row + 1)"
        } else {
            return "\(row + 1)"
        }
    }
    
    func pickerView(
        _ pickerView: UIPickerView,
        didSelectRow row: Int,
        inComponent component: Int
    ) {
        selectedComponent = component
        selectedYear = (self.currentDateCom.year!) - self.pickerView
            .selectedRow(inComponent: 2)
        selectedMonth = self.pickerView.selectedRow(inComponent: 1) + 1
        selectedDay = self.pickerView.selectedRow(inComponent: 0) + 1
        let dateString = String(
            format: "%02ld-%02ld-%02ld",
            selectedDay,
            selectedMonth,
            selectedYear
        )
        model.valueChanged?(dateString)
        if component == 2 {
            let year: Int = pickerView.selectedRow(
                inComponent: 2
            ) + currentDateCom.year!
            let currentMonth = Date().currentMonth
            let currentYear = Date().currentYear
            if year == currentYear{
                months = Array(1...currentMonth)
            }
            pickerView.reloadComponent(1)
        }
        if component == 1 {
            let year: Int = pickerView.selectedRow(
                inComponent: 2
            ) + currentDateCom.year!
            let month: Int = pickerView.selectedRow(inComponent: 1) + 1
            let days: Int = howManyDays(inThisYear: year, withMonth: month)
            let currentMoth = Date().currentMonth
            let currentYear = Date().currentYear
            let currentDay = Date().currentDay
            if year == currentYear,
               month == currentMoth{
                self.days = Array(1 ... currentDay)
            }else{
                self.days = Array(1 ... days)
            }
            pickerView.reloadComponent(0)
        }
    }
}

extension BPDatePickerView{
    func setupUI(){
        backgroundColor = .clear
        addSubview(pickerView)
        NSLayoutConstraint.activate([
            pickerView.leadingAnchor.constraint(equalTo: leadingAnchor),
            pickerView.topAnchor.constraint(equalTo: topAnchor,constant: 44.ratio()),
            pickerView.trailingAnchor.constraint(equalTo: trailingAnchor),
            pickerView.bottomAnchor.constraint(equalTo: bottomAnchor),
            pickerView.heightAnchor.constraint(equalToConstant: 436.ratio()),
        ])
        addSubview(stackView)
        stackView.addArrangedSubview(firstLabel)
        stackView.addArrangedSubview(secondLabel)
        stackView.addArrangedSubview(thirdLabel)
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: leadingAnchor),
            stackView.trailingAnchor.constraint(equalTo: trailingAnchor),
            stackView.topAnchor.constraint(equalTo: topAnchor),
            stackView.heightAnchor.constraint(equalToConstant: 44.ratio())
        ])
    }
    
    func applyModel(){
        self.pickerView.reloadAllComponents()
       let currentDate = Date.stringToDate(self.model.currentDate,dateFormat: "dd-MM-yyyy") ?? Date()
        let dateCom = Calendar.current.dateComponents(
            [.year, .month, .day],
            from: currentDate
        )
        let currentYear = dateCom.year ?? 0
        let currentMonths = dateCom.month ?? 0
        let currentDay = dateCom.day ?? 0
        let currentYearIndex = currentDateCom.year! - currentYear
        self.pickerView
            .selectRow(currentYearIndex, inComponent: 2, animated: true)
        self.pickerView
            .selectRow(currentMonths - 1, inComponent: 1, animated: true)
        self.pickerView
            .selectRow(currentDay - 1, inComponent: 0, animated: true)
        self.pickerView(pickerView, didSelectRow: currentDay - 1, inComponent: 0)
        self.pickerView(pickerView, didSelectRow: currentMonths - 1, inComponent: 1)
        self.pickerView(pickerView, didSelectRow: currentYearIndex, inComponent: 2)
    }
    
    private func howManyDays(inThisYear year: Int, withMonth month: Int) -> Int {
        let daysInMonth = Date.daysInMonth(year: year, month: month)
        return daysInMonth
    }
}

@objc
extension BPDatePickerView{
    @MainActor
    @objc class BPDatePickerViewModel: NSObject {
       @objc var contentHeight = 436.ratio()
      @objc  var currentDate: String = ""
       @objc var valueChanged: ((String)-> Void)? = nil
    }
}
